### ISSUE TITLE
Question about suspicious nonzero behavior with scalar input

### ISSUE THREAD

**Issue by jcwchen** [pytorch/pytorch#76236]
### 🐛 Describe the bug

This nonzero behavior with scalar input looks weird to me:
```
import torch
torch.nonzero(torch.tensor(0))
# output tensor([], size=(0, 0), dtype=torch.int64)
torch.nonzero(torch.tensor(1))
# output tensor([], size=(1, 0), dtype=torch.int64)
```
I thought `torch.nonzero(torch.tensor(1))` should output as `output tensor([0], size=(1, 0), dtype=torch.int64)` especially it has size(1, 0).

Is the original output expected? If so, may I understand more details about why it outputs an empty tensor with size(1, 0)? Thank you for looking into this.

### Versions

PyTorch version: 1.11.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.22.3
Libc version: N/A

Python version: 3.9.7 (default, Sep 16 2021, 16:59:28) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19044-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] numpydoc==1.1.0
[pip3] torch==1.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           haa95532_640
[conda] mkl-service               2.4.0            py39h2bbff1b_0
[conda] mkl_fft                   1.3.1            py39h277e83a_0
[conda] mkl_random                1.2.2            py39hf11a4ad_0
[conda] mypy_extensions           0.4.3            py39haa95532_0
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torchvision               0.12.0                   pypi_0    pypi

cc @ezyang @gchanan @zou3519 @mruberry

**Comment by jcwchen** [comment]
Thank you for looking into this. Since it seems like a wrong behavior, is there any plan to fix this?

**Comment by mruberry** [comment]
Thank you for reporting this issue, @jcwchen -- you're correct the current behavior in PyTorch is incorrect. NumPy shows the correct behavior:

```
np.nonzero(np.array(1))
: (array([0]),)
```

We should probably fix this

**Comment by ezyang** [comment]
Idk the current behavior seems fine. Input has zero dimensions so the output has zero size. Numpy's behavior seems inconsistent with how scalars should be handled.

**Comment by jcwchen** [comment]
Here is the result by NumPy:
```
import numpy as np
np.array(np.nonzero(0))
# array([], shape=(1, 0), dtype=int64)
np.array(np.nonzero(1))
# array([[0]]), shape (1, 1)
```
I think the produced shape by NumPy is very likely incorrect, since they mentioned this behavior for scalar input will be deprecated in the future ([reference](https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html)):
```
When called on a zero-d array or scalar, nonzero(a) is treated as nonzero(atleast_1d(a)).

Deprecated since version 1.17.0: Use [atleast_1d](https://numpy.org/doc/stable/reference/generated/numpy.atleast_1d.html#numpy.atleast_1d) explicitly if this behavior is deliberate.
```
But their result for `np.nonzero(1)`, which is `[0]`, makes more sense to me.

To clarify, I suppose the inferred shape by `torch.nonzero(torch.tensor(1))` is good, but the result is really weird by having an empty tensor with size (1, 0).

**Comment by ezyang** [comment]
0-dim scalar handling is weird, you have to get used to it. torch.nonzero's docs get it exactly right:

> If input has n dimensions, then the resulting indices tensor out is of size (z×n), where z is the total number of non-zero elements in the input tensor. 

The scalar input has 0 dims. There is one nonzero element. So you get 1x0. To do differently would require introducing a special case to the size calculation here. Closing this for now.

**Comment by jcwchen** [comment]
@ezyang Thanks for the explanation. I did check the spec, but it only mentions shape (which is correct) and doesn't specify it will be an empty tensor anyway for shape (0, 0) and (1, 0). I am fine if it is a global case in PyTorch, although it looks weird.

**Comment by ezyang** [comment]
The tensor is empty for those shapes because the numel (product of sizes) is zero ;)
